contract_comm/currency/currency.go

```diff
@@ -216,128 +246,100 @@ func (cc *CurrencyComparator) Cmp(val1 *big.Int, currency1 *common.Address, val2
-		if currency1 != nil {
-			currency1Output = currency1.Hex()
-		}
-		currency2Output := "nil"
-		if currency2 != nil {
-			currency2Output = currency2.Hex()
-		}
-		log.Warn("Error in retrieving exchange rate.  Will do comparison of two values without exchange rate conversion.", "currency1", currency1Output, "err1", err1, "currency2", currency2Output, "err2", err2)
-		return val1.Cmp(val2)
+	celoGoldAddress, err := contract_comm.GetRegisteredAddress(params.GoldTokenRegistryId, nil, nil)
+	if err == errors.ErrSmartContractNotDeployed || err == errors.ErrRegistryContractNotDeployed {
+		return nil, err
 	}
 
-	// Below code block is basically evaluating this comparison:
-	// val1 * exchangeRate1.Denominator/exchangeRate1.Numerator < val2 * exchangeRate2.Denominator/exchangeRate2.Numerator
-	// It will transform that comparison to this, to remove having to deal with fractional values.
-	// val1 * exchangeRate1.Denominator * exchangeRate2.Numerator < val2 * exchangeRate2.Denominator * exchangeRate1.Numerator
-	leftSide := new(big.Int).Mul(val1, new(big.Int).Mul(exchangeRate1.Denominator, exchangeRate2.Numerator))
-	rightSide := new(big.Int).Mul(val2, new(big.Int).Mul(exchangeRate2.Denominator, exchangeRate1.Numerator))
-	return leftSide.Cmp(rightSide)
-}
+	if *currencyAddress == celoGoldAddress {
+		// This function shouldn't really be called with the token's address, but if it is the value
+		// is correct, so return nil as the error
+		return NoopExchangeRate(), nil
+	} else if err != nil {
+		log.Error(err.Error())
+		return nil, err
+	}
 
-func getExchangeRate(currencyAddress *common.Address) (*exchangeRate, error) {
-	var (
-		returnArray [2]*big.Int
-		leftoverGas uint64
-	)
+	var returnArray [2]*big.Int
+	leftoverGas, err := contract_comm.MakeStaticCall(params.SortedOraclesRegistryId, medianRateFuncABI, "medianRate", []interface{}{currencyAddress}, &returnArray, params.MaxGasForMedianRate, header, state)
 
-	if currencyAddress == nil {
-		return &exchangeRate{cgExchangeRateNum, cgExchangeRateDen}, nil
-	} else {
-		if leftoverGas, err := contract_comm.MakeStaticCall(params.SortedOraclesRegistryId, medianRateFuncABI, "medianRate", []interface{}{currencyAddress}, &returnArray, params.MaxGasForMedianRate, nil, nil); err != nil {
-			if err == errors.ErrSmartContractNotDeployed {
-				log.Warn("Registry address lookup failed", "err", err)
-				return &exchangeRate{big.NewInt(1), big.NewInt(1)}, err
-			} else {
-				log.Error("medianRate invocation error", "feeCurrencyAddress", currencyAddress.Hex(), "leftoverGas", leftoverGas, "err", err)
-				return &exchangeRate{big.NewInt(1), big.NewInt(1)}, err
-			}
-		}
+	if err == errors.ErrSmartContractNotDeployed {
+		log.Warn("Registry address lookup failed", "err", err)
+		return NoopExchangeRate(), nil
+	} else if err != nil {
+		log.Error("medianRate invocation error", "feeCurrencyAddress", currencyAddress.Hex(), "leftoverGas", leftoverGas, "err", err)
+		return NoopExchangeRate(), nil
```

Comment:
Oof, this is not great default behavior. We really should not consider the exchange rate of all tokens to be equal to 1/1 when this call fails. We should instead disable any functionality that relies on this. Probably another change we should mark for the next hard fork.

Reply:
Added to the list https://www.notion.so/clabsco/Remove-quirks-list-HardFork-Changes-7ee3a0b904ea464db6af45aba9f94a4a